### PR TITLE
Convert type classes to LF type synonyms

### PR DESCRIPTION
**Pull request by cocreature**
This is still a draft PR until type synonyms are ready across the stack (in particular the Scala side is still missing completely) but the conversion should work and I’m able to run `daml build` with this on a large DAML codebase.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 1]
```diff
@@ -505,16 +520,15 @@ convertBind env (name, x)
...
-        ( (_, tplSuperDict)
-        : (_, key)
-        : _lookupByKey
-        : _fetchByKey
-        : (_, maintainer)
-        : _toAnyContractKey
-        : _fromAnyContractKey
-        : []
-        ) <- convertExpr env x
+      ( templateKeyCon
```
Comment (hurryabit) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]:
```suggestion
       ( templateKeyCon
```

Comment (cocreature) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: :fire:

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 2]
```diff
@@ -1105,8 +1114,11 @@ convertDataCon env m con args
                     [tmArg] -> pure tmArg
                     _ -> unhandled "constructor with more than two unnamed arguments" xargs
 
-            SimpleRecordCon ->
```
Comment (associahedron) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: IMO at this point we should split the handling of simple record types (these remain as `SimpleRecordCon`) and typeclass dictionaries (by adding a new `TypeClassDictCon` or something)

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 3]
```diff
@@ -941,7 +948,9 @@ convertExpr env0 e = do
             let fldIndex = fromJust (elemIndex x vs)
             let fldName = fldNames !! fldIndex
             recTyp <- convertType env (varType bind)
-            pure $ ERecProj (fromTCon recTyp) fldName scrutinee' `ETmApp` EUnit
+            if envLfVersion env `supports` featureTypeSynonyms
```
Comment (associahedron) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: Another suggestion I have is to move (almost) all of these ```envLfVersion env `supports` featureTypeSynonyms``` checks to a set of related utility functions, something like

```Haskell
mkDictProj :: Env -> TypeConApp -> FieldName -> LF.Expr -> LF.Expr
mkDictProj env tcon fieldName
  if envLFVersion env `supports` featureTypeSynonyms
    then EStructProj fieldName
    else ERecProj tcon fieldName
 
mkDictCon :: Env -> TypeConApp -> [(String, LF.Expr)] -> Expr
mkDictcon env tcon =
  if envLFVersion env `supports` featureTypeSynonyms
    then EStructCon
    else ERecCon tcon   
```

etc

**Comment by cocreature** [comment]
@associahedron Both good ideas, I’ve incorporated your suggestions. Thanks!

File: compiler/damlc/tests/daml-test-files/ConstrainedClassMethod.daml [hunk 4]
```diff
@@ -1,5 +1,3 @@
-{-# LANGUAGE ConstrainedClassMethods #-}
```
Comment (cocreature) [inline compiler/damlc/tests/daml-test-files/ConstrainedClassMethod.daml]: This is enabled by default.

File: compiler/damlc/tests/daml-test-files/ConstrainedClassMethod.daml [hunk 5]
```diff
@@ -9,5 +7,8 @@ class A t where
     foo : t -> t
     bar : Eq t => t -> t
 
-class B t where
-    baz : B b => b -> t
+-- This would create a recursive type synonym.
```
Comment (cocreature) [inline compiler/damlc/tests/daml-test-files/ConstrainedClassMethod.daml]: We decided to just disable this for now. We don’t use this on our large internal DAML codebase. There is a reasonable workaround with manual dictionary passing and if it does become a record we could either handle this in the LF conversion or support infinite types (something like OCaml’s `-rectypes`).

File: compiler/damlc/tests/src/unstable-types.sh [hunk 6]
```diff
@@ -32,7 +32,6 @@ for LF_VERSION in $PKG_DB/*; do
         stdlib=$LF_VERSION/daml-stdlib-*.dalf
         prim=$LF_VERSION/daml-prim.dalf
         $DIFF -b -u <(get_serializable_types $stdlib) <(cat <<EOF
-"DA.Upgrade:MetaEquiv"
```
Comment (cocreature) [inline compiler/damlc/tests/src/unstable-types.sh]: This is a typeclass with only a single method which was therefore serializable but is now translated to Unit.

File: compiler/damlc/tests/daml-test-files/ConstrainedClassMethod.daml (hunk 4 above)
Comment (cocreature) [reply compiler/damlc/tests/daml-test-files/ConstrainedClassMethod.daml]: Actually it isn’t. The GHC docs lied to me by claiming that `MultiParamTypeclasses` imply `ConstrainedClassMethods`.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs [hunk 7]
```diff
@@ -944,7 +971,9 @@ convertExpr env0 e = do
...
+              then pure $ EStructProj fldName scrutinee' `ETmApp` EUnit
+              else pure $ ERecProj (fromTCon recTyp) fldName scrutinee' `ETmApp` EUnit
```
Comment (associahedron) [inline compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: I think you can replace this `if` expression by using the helper function,

```Haskell
pure $ mkDictProj env (fromTCon recTyp) fldName scrutinee' `ETmApp` EUnit
```

File: compiler/damlc/tests/daml-test-files/ConstrainedClassMethod.daml (hunk 5 above)
Comment (associahedron) [reply compiler/damlc/tests/daml-test-files/ConstrainedClassMethod.daml]: Seems reasonable to me. I don't know of any natural examples of a class that requires itself as a constraint.

File: compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs (hunk 7 above)
Comment (cocreature) [reply compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs]: good catch, fixed. Thank you!

**Pull request merged**